### PR TITLE
Email chat notifications

### PR DESCRIPTION
**Pull request by shirishgoyal**
Email Notifications
- When user is offline or hasn't checked messages for last 30 minutes, messages are sent offline via digest email
- Emails are not sent for the same messages again
- Handle message status for sent, delivered and read
- Handle state transition from login to previous state if user is not authenticated

Refactoring
- Project

### CODE REVIEW

File: crowdsourcing/tasks.py [hunk 1]
```diff
@@ -54,3 +62,52 @@ def update_worker_cache(workers, operation, key=None, value=None):
...
+                result[message_recipient['message__sender__username']] = [message_recipient]
+
+        messages = [{'sender': k, 'messages': v} for k, v in result.items()]
+
+        if len(messages) > 0:
+            # send email
+            send_notifications_email(email=user.email, url=url, messages=messages)
+
+            # update Email Notification for user - updated_at
+            models.EmailNotification.objects.filter(recipient=user).update(updated_at=timezone.now())
```
Comment (dmorina) [inline crowdsourcing/tasks.py]: let's move this out of the loop

File: crowdsourcing/viewsets/project.py [hunk 2]
```diff
@@ -48,61 +15,66 @@ class ProjectViewSet(viewsets.ModelViewSet):
     permission_classes = [IsProjectOwnerOrCollaborator, IsAuthenticated]
 
     def create(self, request, with_defaults=True, *args, **kwargs):
-        project_serializer = ProjectSerializer(data=request.data, fields=('name', 'price', 'post_mturk', 'repetition',
-                                                                          'template'))
-        if project_serializer.is_valid():
+        serializer = ProjectSerializer(
+            data=request.data,
+            fields=('name', 'price', 'post_mturk', 'repetition', 'templates'),
```
Comment (dmorina) [inline crowdsourcing/viewsets/project.py]: templates -> template

File: crowdsourcing/tasks.py [hunk 3]
```diff
@@ -107,7 +107,9 @@ def email_notifications():
             # send email
             send_notifications_email(email=user.email, url=url, messages=messages)
 
-            # update Email Notification for user - updated_at
-            models.EmailNotification.objects.filter(recipient=user).update(updated_at=timezone.now())
+            users_notified.append(user)
+
+        # update the last time user was notified
+        models.EmailNotification.objects.filter(recipient__in=users_notified).update(updated_at=timezone.now())
```
Comment (dmorina) [inline crowdsourcing/tasks.py]: unindent this

File: static/js/authentication/controllers/login.controller.js [hunk 4]
```diff
@@ -73,7 +74,17 @@
                 }).finally(function () {
                 });
             }
-            vm.submitted=true;
+            vm.submitted = true;
+        }
+
+        function gotoNext() {
+            var next = $stateParams.next;
```
Comment (dmorina) [inline static/js/authentication/controllers/login.controller.js]: let's capture the next params as well  example url `/task/32`

Comment (shirishgoyal) [inline static/js/authentication/controllers/login.controller.js]: Done

File: crowdsourcing/serializers/project.py [hunk 5]
```diff
@@ -83,19 +71,49 @@ def create(self, with_defaults=True, **kwargs):
...
-        # revision.repetition = new_data.get('repetition', revision.repetition)
-        # revision.price = new_data.get('repetition', revision.price)
-        # revision.deadline = new_data.get('repetition', revision.deadline)
-        revision.published_time = timezone.now()
-        revision.save()
+    def update(self, *args, **kwargs):
+        status = self.validated_data.get('status', self.instance.status)
+        num_rows = self.validated_data.get('num_rows', 0)
+        if self.instance.status != status and status == 2:
+            if self.instance.templates.all()[0].items.count() == 0:
```
Comment (dmorina) [inline crowdsourcing/serializers/project.py]: `template`

File: crowdsourcing/viewsets/project.py [hunk 6]
```diff
@@ -48,61 +15,66 @@ class ProjectViewSet(viewsets.ModelViewSet):
...
+            return Response(data=serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
     def retrieve(self, request, *args, **kwargs):
-        project_object = self.get_object()
-        serializer = ProjectSerializer(instance=project_object,
+        serializer = ProjectSerializer(instance=self.get_object(),
                                        fields=('id', 'name', 'price', 'repetition', 'deadline', 'timeout',
-                                               'is_prototype', 'template', 'status', 'batch_files', 'post_mturk',
-                                               'qualification'))
+                                               'is_prototype', 'templates', 'status', 'batch_files', 'post_mturk'
```
Comment (dmorina) [inline crowdsourcing/viewsets/project.py]: `template`

**Pull request merged**